### PR TITLE
[FIX] web_editor: get image source while modifying it

### PR DESCRIPTION
**Pull request by sami-odoo**
If the user is using the `note` or any other module to add and modify images,
and they try to crop the image and apply the changes, they may face an error if
they try to crop it again without clicking on `apply changes` and instead click
somewhere outside the image.

steps to produce:
- Install `note`.
- Create a note, add an image in it.
- Click on image to crop it, select the portion you want to crop and then
  click on `Apply` to apply changes.
- Select the image again, click on crop icon and then click anywhere outside the
  image boundaries.

Traceback:
```
KeyError: 667
  File "odoo/api.py", line 959, in get
    cache_value = field_cache[record._ids[0]]
CacheMiss: 'ir.attachment(667,).image_src'
  File "odoo/fields.py", line 1158, in __get__
    value = env.cache.get(record, self)
  File "odoo/api.py", line 966, in get
    raise CacheMiss(record, field)
TypeError: 'bool' object is not subscriptable
  File "odoo/http.py", line 2123, in __call__
    response = request._serve_db()
  File "odoo/http.py", line 1699, in _serve_db
    return service_model.retrying(self._serve_ir_http, self.env)
  File "odoo/service/model.py", line 133, in retrying
    result = func()
  File "odoo/http.py", line 1726, in _serve_ir_http
    response = self.dispatcher.dispatch(rule.endpoint, args)
  File "odoo/http.py", line 1927, in dispatch
    result = self.request.registry['ir.http']._dispatch(endpoint)
  File "addons/website/models/ir_http.py", line 234, in _dispatch
    response = super()._dispatch(endpoint)
  File "odoo/addons/base/models/ir_http.py", line 190, in _dispatch
    result = endpoint(**request.params)
  File "odoo/http.py", line 716, in route_wrapper
    result = endpoint(self, *args, **params_ok)
  File "addons/web_editor/controllers/main.py", line 577, in modify_image
    return attachment.image_src
  File "odoo/fields.py", line 1209, in __get__
    self.compute_value(recs)
  File "odoo/fields.py", line 1387, in compute_value
    records._compute_field_value(self)
  File "odoo/models.py", line 4486, in _compute_field_value
    fields.determine(field.compute, self)
  File "odoo/fields.py", line 99, in determine
    return needle(*args)
  File "addons/web_editor/models/ir_attachment.py", line 47, in _compute_image_src
    unique = attachment.checksum[:8]
```

Applying this commit will solve the issue because, after applying it, whenever
`modify_image` is called, `data` will receive the proper source.

sentry-4185609374

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by sami-odoo** [comment]
Hello @odoo/sentry_reviewers ,
Please review it.

**Comment by mart-e** [comment]
>  and if the attachment.checksum results to be false or if the checksum field of the ir.attachment is false

and why is that? Seems like hidding the symptoms instead of the cause

File: addons/web_editor/models/ir_attachment.py [hunk 1]
```diff
@@ -36,7 +36,7 @@ def _compute_image_src(self):
                 attachment.image_src = attachment.url
             else:
                 # Adding unique in URLs for cache-control
-                unique = attachment.checksum[:8]
+                unique = (attachment.checksum or '')[:8]
```
Comment (mart-e) [inline addons/web_editor/models/ir_attachment.py:39]: so if unique is not defined, it should not be in the URL. You don't want a `foo.jpg?unique=` or `/we/image/42-/foo`

**Comment by mart-e** [comment]
> The checksum might be empty when attachment record has type url

But the code is
```py
if attachment.type == 'url':
    ...
else:
    unique = attachment.checksum[:8]
```
so your image is _not_ an URL when you have the error. Or you are in a strange situation where `type == 'binary' and no checksum. Shouldn't this one be fixed instead?

**Comment by sami-odoo** [comment]
Hello @yelizariev ,
The solution was provided according to the function
https://github.com/odoo/odoo/blob/ee680a9365bba0ed4c0e5e741b3e8dce94fab9bd/addons/web_editor/models/ir_attachment.py#L30

where we are only calculating checksum of attachment which are not in url form.

**Comment by abz-odoo** [comment]
Hello @sami-odoo i think you didnt give your feedback to Martin

**Comment by sami-odoo** [comment]
Hello @mart-e 
Sorry for a late reply!

There are some changes i have made in this PR.
- I have added steps to produce.
- Changed the code .
- Retargetted the pr to v 16.0

Whenever `modify_image` was being executed
https://github.com/odoo/odoo/blob/721b996685c8f6943832acfce52b05c8773522f9/addons/web_editor/static/src/js/wysiwyg/wysiwyg.js#L980 

`data` always had its value evaluted to None.

Because,
When it is image 
https://github.com/odoo/odoo/blob/721b996685c8f6943832acfce52b05c8773522f9/addons/web_editor/static/src/js/wysiwyg/wysiwyg.js#L971

`isBackground` is `False`
And,
 
> el.getAttribute('src')

Whenever we are modifying or cropping the image for the first time 
`src` is always in the form below:
- `data: img/(jpg/png/jpeg);base64, (some base64 encoded characters)` . So whenever we used .split(',')[1]
we always got base64 encoded characters.

But, whenever we are selecting and modifying image for the second time 
`src` is in the form
For eg: `/web/image/208-5cdf6eaf/icon.png` and here nothing is comma seperated. Because of which `data` was none
and users were facing issue.

**Comment by mart-e** [comment]
Thanks for the update, as it's no longer modifying base, I let the editor team review this PR :)

**Comment by dmo-odoo** [comment]
This looks to be due to a renaming of a class from one version to another. Looks legit enough. However, I'm confused by the fact that this PR targets 16.0 but the branch name starts with 14.0. Are you sure your intended target is 16.0 and therefore that means the class has changed from 15.0 to 16.0 ? Just making sure.

**Comment by sami-odoo** [comment]
Hello @dmo-odoo ,
> However, I'm confused by the fact that this PR targets 16.0 but the branch name starts with 14.0. Are you sure your intended target is 16.0 

- The issue is reproducable in v16 and it's later versions. 

- Initially the code was to handle the traceback so we gave the fix from v14, but now when we got the proper cause of the issue we intend to solve it from the version where it's produced.

**Comment by dmo-odoo** [comment]
@robodoo r+

**Pull request closed without merging**